### ISSUE TITLE
53612: bind: address already in use

### ISSUE THREAD

**Issue by JachMinen** [aws-observability/aws-otel-lambda#808]
---
name: Bug report
about: Create a report to help us improve
title: ''
labels: bug
assignees: ''

---

**Describe the bug**
Lambda will occasionally encounter this error during initialization of environment:
`telemetryAPI.Listener`
`Unexpected stop on HTTP Server`
`listen tcp ---.---.---.---:53612: bind: address already in use`
When this happens, the next invocation will run successfully at first, but then time out when trying to export metrics.
This only seems to happen on cold start of new environment and usually resolves itself after 1 or 2 invocations in that environment.

**Steps to reproduce**
Unknown

**What did you expect to see?**
Lambda runs and metrics are exported successfully

**What did you see instead?**
Lambda times out

**What version of collector/language SDK version did you use?**
Version: aws-otel-python-amd64-ver-1-20-0:3

**What language layer did you use?**
Config: Python

**Additional context**
Lambda also has layers for Lambda Insights and Python libraries

**Comment by drobert** [comment]
We've been bumping into this, too. We're using a fastapi python app in a docker-based lambda also using `aws-otel-python-amd64-ver-1-20-0:3`

**Comment by Xaaris** [comment]
Same here on `aws-otel-collector-arm64-ver-0-90-1:1`

**Comment by jan-xyz** [comment]
@drobert and @JachMinen are you using `otlptracegrpc.WithDialOption(grpc.WithBlock())` for your SDK client setup?

**Comment by drobert** [comment]
Please keep this open. Unresolved/still an issue.

**Comment by Joe-Woodhead-BigChange** [comment]
We encountered this today. Happened twice in a row on deployment to Production where it hit a timeout on cold start. This is a .NET 8 lambda using this layer pushing data into New Relic:

_arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-collector-amd64-ver-0-102-1:1_

Error below:

{
    "level": "error",
    "ts": 1729787819.487711,
    "logger": "telemetryAPI.Listener",
    "msg": "Unexpected stop on HTTP Server",
    "error": "listen tcp 169.254.79.130:53612: bind: address already in use"
}

**Comment by artiom** [comment]
Still encountering this issue.

```
Python 3.12 Lambda
OTEL layer: arn:aws:lambda:AWS_REGION:901920570463:layer:aws-otel-python-amd64-ver-1-29-0:1
```

```
INIT_START Runtime Version: python:3.12.v38 Runtime Version ARN: arn:aws:lambda:eu-central-1::runtime:***
{"level":"info","ts":1737796522.1585064,"msg":"Launching OpenTelemetry Lambda extension","version":"v0.115.0"}
{"level":"info","ts":1737796522.1701703,"logger":"telemetryAPI.Listener","msg":"Listening for requests","address":"sandbox.localdomain:53612"}
{"level":"info","ts":1737796522.170302,"logger":"telemetryAPI.Client","msg":"Subscribing","baseURL":"http://127.0.0.1:9001/2022-07-01/telemetry"}
TELEMETRY Name: collector State: Subscribed Types: [Platform]
{"level":"error","ts":1737796522.1730578,"logger":"telemetryAPI.Listener","msg":"Unexpected stop on HTTP Server","error":"listen tcp 169.254.79.130:53612: bind: address already in use"}
{"level":"info","ts":1737796522.1736264,"logger":"telemetryAPI.Client","msg":"Subscription success","response":"\"OK\""}
{"level":"info","ts":1737796522.1736798,"logger":"NewCollector","msg":"Using config URI from environment variable","uri":"/opt/python/observability/metrics/opentelemetry_collector.yaml"}
{"level":"info","ts":1737796522.6072795,"caller":"service@v0.115.0/service.go:166","msg":"Setting up own telemetry..."}
{"level":"info","ts":1737796522.6074731,"caller":"telemetry/metrics.go:70","msg":"Serving metrics","address":"localhost:8888","metrics level":"Normal"}
{"level":"info","ts":1737796522.6100013,"caller":"service@v0.115.0/service.go:238","msg":"Starting aws-otel-lambda...","Version":"v0.115.0","NumCPU":2}
{"level":"info","ts":1737796522.6100285,"caller":"extensions/extensions.go:39","msg":"Starting extensions..."}
{"level":"info","ts":1737796522.610041,"caller":"extensions/extensions.go:42","msg":"Extension is starting...","kind":"extension","name":"sigv4auth"}
{"level":"info","ts":1737796522.610055,"caller":"extensions/extensions.go:59","msg":"Extension started.","kind":"extension","name":"sigv4auth"}
{"level":"info","ts":1737796522.6273103,"caller":"otlpreceiver@v0.115.0/otlp.go:169","msg":"Starting HTTP server","kind":"receiver","name":"otlp","data_type":"metrics","endpoint":"localhost:4318"}
{"level":"info","ts":1737796522.627401,"caller":"service@v0.115.0/service.go:261","msg":"Everything is ready. Begin running and processing data."}
```

**Comment by Dreamescaper** [comment]
Does anyone here has this issue if using _without_ AWS Lambda Insights?

**Comment by jan-xyz** [comment]
yes the adot layer is the only one we have attached and we have this error

**Comment by Dreamescaper** [comment]
Apparently, this issue should be fixed in 0-115-0:2

**Comment by Dreamescaper** [comment]
My bad.
AWS Support suggested to update to this version, and I thought for a moment that this layer was released today. 
Yeah, we still have this issue as well.

**Comment by crblack82** [comment]
My aws support contact suggested updating to:

arn:aws:lambda:<region>:901920570463:layer:aws-otel-python-<architecture>-ver-1-29-0:2

where I was previously seeing the error when using:

arn:aws:lambda:<region>:901920570463:layer:aws-otel-python-<architecture>-ver-1-29-0:1

I can confirm that after deploying the change to 40+ lambdas, many of which run every one minute, we have a 100% drop off rate after 3 days in terms of the error "bind: address already in use". I will be responding to my case stating that the issue does appear fixed.

**Comment by vasireddy99** [comment]
This Issue is fixed and new patched layers are released - https://aws-otel.github.io/docs/ReleaseBlogs/aws-distro-for-opentelemetry-lambda-layer-patched-v0.42.0

Thanks @crblack82
